mapper-processor/src/main/java/com/datastax/oss/driver/internal/mapper/processor/util/HierarchyScanner.java

```diff
@@ -0,0 +1,241 @@
+   *   <li>Visit <code>typeElements</code>'s interfaces, and record those interfaces' parents for
+   *       later use (<code>newInterfacesToScan</code>)
+   *   <li>Visit <code>interfacesToScan</code>, and append those interface's parents to <code>
+   *       newInterfacesToScan</code> for later use)
+   *   <li>If <code>superClassElement != null</code> </code>Set <code>typeElement :=
+   *       superClassElement, interfacesToScan := newInterfacesToScan</code> and repeat starting at
+   *       step 3
+   *   <li>Visit <code>newInterfacesToScan</code> interfaces and their parents until we've reached
+   *       root
+   * </ol>
+   *
+   * Once a {@link HierarchyScanStrategy} is identified, the returning hierarchy is built by
+   * traversing the hierarchy again using the chosen strategy.
+   *
+   * @param typeElement The type whose hierarchy will be traversed.
+   * @param context provides utilities for working with types.
+   * @return The type hierarchy, ordered from typeElement to the highestAncestor, as dictated by the
+   *     resolved {@link HierarchyScanStrategy}
+   */
+  public static Set<TypeMirror> resolveTypeHierarchy(
+      TypeElement typeElement, ProcessorContext context) {
+    HierarchyScanStrategyOptions hierarchyScanStrategy =
+        HierarchyScanner.resolveHierarchyScanStrategy(typeElement, context);
+
+    ImmutableSet.Builder<TypeMirror> hierarchy = ImmutableSet.builder();
+    traverseFullHierarchy(hierarchyScanStrategy, typeElement, context, hierarchy::add);
+    return hierarchy.build();
+  }
+
+  private static HierarchyScanStrategyOptions resolveHierarchyScanStrategy(
+      TypeElement classElement, ProcessorContext context) {
+    // Use the default HierarchyScanStrategy to find the configured HierarchyScanStrategy.
+    // This is done because the default strategy is the most permissive.
+
+    // traverse hierarchy until a strategy is found.
+    final HierarchyScanStrategyOptions defaultOptions =
+        new HierarchyScanStrategyOptions(defaultStrategy);
+    final AtomicReference<HierarchyScanStrategyOptions> ref = new AtomicReference<>(defaultOptions);
+    traverseHierarchy(
+        defaultOptions,
+        classElement,
+        context,
+        (TypeMirror mirror) -> {
+          // if we find a strategy, set it and stop traversing.
+          Element element = context.getTypeUtils().asElement(mirror);
+          for (AnnotationMirror candidate : element.getAnnotationMirrors()) {
+            if (context
+                .getClassUtils()
+                .isSame(candidate.getAnnotationType(), HierarchyScanStrategy.class)) {
+              ref.compareAndSet(defaultOptions, new HierarchyScanStrategyOptions(candidate));
+              return false;
+            }
+          }
+          return true;
+        });
+
+    return ref.get();
+  }
+
+  private static class HierarchyScanStrategyOptions {
+    boolean scanAncestors = defaultStrategy.scanAncestors();
+    TypeMirror highestAncestor = null;
```

Comment:
why we don't take `defaultStrategy.highestAncestor`?

Reply:
Good catch.  I think it ultimately doesn't have an effect, but we can get a TypeMirror for that (Object.class) :+1: